### PR TITLE
mumps: new version and OpenMP support

### PR DESCRIPTION
**Pull request by lpoirel**

### CODE REVIEW

**Comment by lpoirel** [comment]
Bonjour @RemiLacroix-IDRIS,
You worked on this package recently, would you mind reviewing these changes ?

**Comment by RemiLacroix-IDRIS** [comment]
Hi,

Looks mostly good to me, just two things:
 - I'm not sure whether OpenMP should be enabled by default or not. I happen to have been working on a PR of my own (#23161) but I lacked time to upstream it. It has OpenMP disabled by default just to stay on the safe side, being that I am not sure if all packages depending on MUMPS will want an OpenMP version of it. It's probably not that crucial though. Someone else might have some thoughts about this.
 - I think I would keep building the examples. If you just do `make` it will build `dexamples`.

**Comment by RemiLacroix-IDRIS** [comment]
> I think I would keep building the examples. If you just do make it will build dexamples.

Scratch that, the `dexamples` target doesn't exist anymore in version 5.4.0, it has basically replaced the old `d` target.

If we want to keep a consistent behavior, versions `@:5.3.5` should keep adding `examples`.

**Comment by lpoirel** [comment]
Regarding the default for the openmp variant in mumps, I was not sure either.
I did a little grep to see what was the common practice: nearly half the packages with an "openmp" variant default it to `True`, nearly half default it to `False`, the remaining defaulting it to `sys.platform != 'darwin'`.
I am happy with writing `spack install mumps +openmp` if it is not the default, but I think it should be: openmp parallelism is now a quite standard feature of mumps, that one could expect to work out of the box with `spack install mumps`.
I understand that this could break packages that depend on mumps: I reviewed them to propagate the option when it exists.

@keitat is my change relevant to trilinos?

**Comment by lpoirel** [comment]
@RemiLacroix-IDRIS  I can also drop the openmp stuff from this PR, since it is already covered by #23161 or I could give you access to my branch so that we merge the two PR?

**Comment by RemiLacroix-IDRIS** [comment]
@lpoirel: My PR can wait that yours is merged, that's not a problem.

@balay @BarrySmith @jedbrown: does this matter for PETSc?

**Comment by jedbrown** [comment]
PETSc 'main' supports mumps-5.4.0, and the commit to make that change did not require any source changes (only how PETSc `--download-mumps` runs the MUMPS build system) so it should also work with petsc-3.15 (though I haven't tested that).

As for OpenMP, the only concern would be OpenMP-unaware apps oversubscribing in calling PETSc (for example) with many ranks per node, and then having OpenMP spawn lots of threads per rank. That's unfortunate complexity leakage, but not relevant at this packaging stage.

**Comment by RemiLacroix-IDRIS** [comment]
@lpoirel : Sorry there has been some kind of mix-up, my PR has been merged before yours. That's my bad, I didn't update my PR to make my plan clear.

It doesn't really matter though, you can make OpenMP enabled default in yours. I shouldn't be too painful to rebase.

File: var/spack/repos/builtin/packages/mumps/package.py
Comment (lpoirel) [inline var/spack/repos/builtin/packages/mumps/package.py:330]: @RemiLacroix-IDRIS is this OK?

Comment (RemiLacroix-IDRIS) [reply var/spack/repos/builtin/packages/mumps/package.py:330]: Looks good to me!

**Comment by alalazo** [comment]
@sethrj Can you check if the modifications on `trilinos` are ok?

File: var/spack/repos/builtin/packages/mumps/package.py
Comment (lpoirel) [inline var/spack/repos/builtin/packages/mumps/package.py:326]: `# Build fails in parallel`
This is no longer true.

The change (no more `make dexamples`) is due to some work on the makefiles to allow parallel build (`make -j all`) according to Mumps developers.

**Review by sethrj — APPROVED** [review]
Looks good to me

File: var/spack/repos/builtin/packages/mumps/package.py
Comment (RemiLacroix-IDRIS) [reply var/spack/repos/builtin/packages/mumps/package.py:326]: Yeah but I think it is simpler to keep it the way it currently is to support older versions.

**Comment by lpoirel** [comment]
@alalazo you asked @sethrj for a review for the trilinos changes, and he approved my PR. 
@keitat [spack trilinos maintainer] did not respond in this PR but mentioned it in trilinos/Trilinos/issues/9042:
"In addition to that, MUMPS team is making some modification in Spack script to introduce OpenMP. I do not think this will affect Spack-Trilinos. I am going to investigate further.
See spack/spack#23143 (comment)"

What should we do so that mumps@5.4.0 is not blocked in spack while @keitat is investigating my PR further? 
- Is @sethrj 's approval sufficient in the lack of  a comment here by @keitat? 
- Should I remove my https://github.com/spack/spack/pull/23143/commits/fa8ba2e5dbcdcee9e2f9ff8a6f57be4f697fbcd2 commit?
- Should I revert mumps in the trilinos package to the old default `depends_on('mumps@5.0:+mpi+shared~openmp', when='+mumps')`

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@sethrj 's approval is sufficient for me. If there's any issue raised after this merge we can work it out in a following PR

**Comment by lpoirel** [comment]
Thanks

**Comment by keitat** [comment]
Just building the new MUMPS with Amesos2.  I did not see any problem.

**Pull request merged**
